### PR TITLE
[Filesystem] Add the `readFile()` method

### LINKED ISSUE #47215: RFC: FileSystem::getContent
### Description

Currently the file system allows to move/create/remove file but also to write content or to append content in a file.
But there is no method to get the method from the file.

Even if the implementation of the method might be easy, what do you think about adding a `getContent(string $name)` method in order to allow doing every content/file manipulation with the symfony methods.

This could also be helpful in test, allowing to mock the `file_get_content` method.

### Example

_No response_

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #47215
| License       | MIT

Our `Filesystem` class is a tool for leveraging PHP's native filesystem functions is a sane way.

For example, it provides a method `dumpFile()` which attempts to write the contents of a string to a file and throws if that attempt fails. However, it lacks a method to read the content of a file. The native function for this task, `file_get_contents()` comes with a couple of traps that I don't want to think about or step into when working with a file system:

* If reading a file fails, `file_get_contents()` returns `false` and triggers an `E_WARNING`. I need to check the return value against `false` and if I want feedback on the failure, I need to work with custom error handlers. Yuck.
* Passing the path to a directory to `file_get_contents()` is almost certainly a mistake that I would want to get an exception for. However, `file_get_contents()` will return an empty string in that case. Fun times.

With this PR, I'm proposing to add a `readFile()` method that either returns the content of a given file or throws an `IOException`. Furthermore, I'm dogfooding this method in components that already depend on the Filesystem component.

### CODE REVIEW

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 1]
```diff
@@ -178,7 +178,7 @@ public function testCopyForOriginUrlsAndExistingLocalFileDefaultsToCopy()
         $this->filesystem->copy($sourceFilePath, $targetFilePath, false);
 
         $this->assertFileExists($targetFilePath);
-        $this->assertEquals(file_get_contents($sourceFilePath), file_get_contents($targetFilePath));
+        $this->assertEquals(file_get_contents($sourceFilePath), file_get_contents($targetFilePath));;
```
Comment (andrew-demb) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php]:
```suggestion
        $this->assertEquals(file_get_contents($sourceFilePath), file_get_contents($targetFilePath));
```

File: src/Symfony/Component/AssetMapper/Tests/Command/AssetMapperCompileCommandTest.php [hunk 2]
```diff
@@ -48,10 +48,9 @@ public function testAssetsAreCompiled()
             $this->filesystem->remove($targetBuildDir);
         }
         // put old "built" versions to make sure the system skips using these
-        $this->filesystem->mkdir($targetBuildDir);
-        file_put_contents($targetBuildDir.'/manifest.json', '{}');
-        file_put_contents($targetBuildDir.'/importmap.json', '{}');
-        file_put_contents($targetBuildDir.'/entrypoint.file6.json', '[]');
+        $this->filesystem->dumpFile($targetBuildDir.'/manifest.json', '{}');
```
Comment (andrew-demb) [inline src/Symfony/Component/AssetMapper/Tests/Command/AssetMapperCompileCommandTest.php:51]: There was a call for mkdir - is it necessary to remain?

```suggestion
        $this->filesystem->mkdir($targetBuildDir);
        $this->filesystem->dumpFile($targetBuildDir.'/manifest.json', '{}');
```

Comment (derrabus) [reply src/Symfony/Component/AssetMapper/Tests/Command/AssetMapperCompileCommandTest.php:51]: `dumpFile()` should already make sure, the directory we're dumping our file into exists.

File: src/Symfony/Component/Filesystem/Filesystem.php [hunk 3]
```diff
@@ -694,6 +694,25 @@ public function appendToFile(string $filename, $content, bool $lock = false): vo
...
+    /**
+     * Returns the content of a file as a string.
+     *
+     * @throws IOException If the file cannot be read
+     */
+    public function readFile(string $filename): string
+    {
+        if (is_dir($filename)) {
+            throw new IOException(sprintf('Failed to read file "%s": File is a directory.', $filename));
+        }
```
Comment (jprivet-dev) [inline src/Symfony/Component/Filesystem/Filesystem.php:706]: Hi! The idea and implementation are great :smiley: 

Would there be any point, in `src/Symfony/Component/Filesystem/Filesystem::readFile`, in checking whether the file is readable, as in [`src/Symfony/Component/Config/Util/XmlUtils::loadFile()`](https://github.com/symfony/symfony/blob/0019010afabb189ecf220c5c05c4b1e45d615acf/src/Symfony/Component/Config/Util/XmlUtils.php#L122)?

```suggestion
        if (is_dir($filename)) {
            throw new IOException(sprintf('Failed to read file "%s": File is a directory.', $filename));
        }
        
        if (is_readable($filename)) {
            throw new IOException(sprintf('Failed to read file "%s": File is not readable.', $filename));
        }
```

Comment (derrabus) [reply src/Symfony/Component/Filesystem/Filesystem.php:706]: If the file is not readable (for whatever reason), `file_get_contents()` triggers a meaningful warning that my implementation escalates to an exception already.

Comment (jprivet-dev) [reply src/Symfony/Component/Filesystem/Filesystem.php:706]: Indeed  !

To test the idea to the limit (having a specific message if the file is not readable), I did a simple test:

```
-rw-r--r-- 1 root root   23 Mar  8 00:00 readable.md
--w------- 1 root root   23 Mar  8 00:00 not-readable.md
```

With `not-readable.md`, we have the following exception with the current version of the implementation:

```
IOException

Failed to read file "/.../not-readable.md": file_get_contents(/.../not-readable.md): Failed to open stream: Permission denied
```

If I force an exception with `is_readable()` in the `readFile()`, I get some erratic behaviour. I have an exception at `clear:cache`:

```
$ bin/console cache:clear

In Filesystem.php line 709:                                                                                                                                                                        
Failed to read file "/.../var/cache/dev/App_KernelDevDebugContainer.xml.meta": File is not readable.  
```

Or with `link`:

```
$ php link ../pr_54173
PHP Fatal error:  Uncaught Symfony\Component\Filesystem\Exception\IOException: Failed to read file "/.../symfony/src/Symfony/Bundle/TwigBundle/composer.json": File is not readable. in /.../symfony/src/Symfony/Component/Filesystem/Filesystem.php:709
```

On https://www.php.net/manual/en/function.is-readable.php, it is noted that:

> The results of this function are cached. See clearstatcache() for more details.

In the `readFile()` method, I run `clearstatcache()` before, but still get the same exceptions :-1: 

Conclusion: in the `readFile()` method, everything works fine without `is_readable()` :+1:

File: src/Symfony/Component/AssetMapper/CompiledAssetMapperConfigReader.php [hunk 4]
```diff
@@ -11,15 +11,19 @@
...
 use Symfony\Component\Filesystem\Path;
 
 /**
  * Reads and writes compiled configuration files for asset mapper.
  */
 class CompiledAssetMapperConfigReader
 {
+    private readonly Filesystem $filesystem;
+
     public function __construct(private readonly string $directory)
```
Comment (OskarStark) [inline src/Symfony/Component/AssetMapper/CompiledAssetMapperConfigReader.php]: Could you please linebreak while on it? Thanks

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/AssetMapper/CompiledAssetMapperConfigReader.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/AssetMapper/CompiledAssetMapperConfigReader.php]: Sure.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @derrabus.

**Pull request merged**